src/shelly_hap_rgb.cpp

```diff
@@ -101,113 +103,126 @@ Status RGB::Init() {
-      std::bind(&RGB::HandleSaturationRead, this, _1, _2, _3),
+      std::bind(&RGBWLight::HandleSaturationRead, this, _1, _2, _3),
       true /* supports_notification */,
-      std::bind(&RGB::HandleSaturationWrite, this, _1, _2, _3),
+      std::bind(&RGBWLight::HandleSaturationWrite, this, _1, _2, _3),
       kHAPCharacteristicDebugDescription_Saturation);
   state_notify_chars_.push_back(saturation_char);
   AddChar(saturation_char);
 
   return Status::OK();
 }
 
-void RGB::HSVtoRGB(float h, float s, float v, float &r, float &g, float &b) {
-  if (s == 0.0) {
+void RGBWLight::HSVtoRGBW(const HSV &hsv, RGBW &rgbw) const {
+  if (hsv.s == 0.0) {
     // if saturation is zero than all rgb channels same as brightness
-    r = g = b = v;
-    return;
-  }
+    rgbw.r = rgbw.g = rgbw.b = hsv.v;
+  } else {
+    // otherwise calc rgb from hsv
+    int i = static_cast<int>(hsv.h * 6);
+    float f = (hsv.h * 6.0f - i);
+    float p = hsv.v * (1.0f - hsv.s);
+    float q = hsv.v * (1.0f - f * hsv.s);
+    float t = hsv.v * (1.0f - (1.0f - f) * hsv.s);
+
+    switch (i % 6) {
+      case 0:
+        rgbw.r = hsv.v;
+        rgbw.g = t;
+        rgbw.b = p;
+        break;
 
-  int i = static_cast<int>(h * 6);
-  float f = (h * 6.0f - i);
-  float p = v * (1.0f - s);
-  float q = v * (1.0f - f * s);
-  float t = v * (1.0f - (1.0f - f) * s);
-
-  switch (i % 6) {
-    case 0:
-      r = v;
-      g = t;
-      b = p;
-      break;
+      case 1:
+        rgbw.r = q;
+        rgbw.g = hsv.v;
+        rgbw.b = p;
+        break;
 
-    case 1:
-      r = q;
-      g = v;
-      b = p;
-      break;
+      case 2:
+        rgbw.r = p;
+        rgbw.g = hsv.v;
+        rgbw.b = t;
+        break;
 
-    case 2:
-      r = p;
-      g = v;
-      b = t;
-      break;
+      case 3:
+        rgbw.r = p;
+        rgbw.g = q;
+        rgbw.b = hsv.v;
+        break;
 
-    case 3:
-      r = p;
-      g = q;
-      b = v;
-      break;
+      case 4:
+        rgbw.r = t;
+        rgbw.g = p;
+        rgbw.b = hsv.v;
+        break;
 
-    case 4:
-      r = t;
-      g = p;
-      b = v;
-      break;
+      case 5:
+        rgbw.r = hsv.v;
+        rgbw.g = p;
+        rgbw.b = q;
+        break;
+    }
+  }
 
-    case 5:
-      r = v;
-      g = p;
-      b = q;
-      break;
+  if (GetMode() == Mode::kRgbw) {
+    // apply white channel to rgb if activated
+    rgbw.w = std::min(rgbw.r, std::min(rgbw.g, rgbw.b));
+    rgbw.r = rgbw.r - rgbw.w;
+    rgbw.g = rgbw.g - rgbw.w;
+    rgbw.b = rgbw.b - rgbw.w;
+  } else {
+    // otherwise turn white channel off
+    rgbw.w = 0.0f;
   }
 }
 
-void RGB::SetOutputState(const char *source) {
+void RGBWLight::SetOutputState(const char *source) {
   LOG(LL_INFO,
       ("state: %s, brightness: %i, hue: %i, saturation: %i", OnOff(cfg_->state),
        cfg_->brightness, cfg_->hue, cfg_->saturation));
 
-  float r = 0, g = 0, b = 0;
+  RGBW rgbw{0.0f, 0.0f, 0.0f, 0.0f};
```

Comment:
`RGBW rgbw{}`;

avoid aggregate initalizers, they are silent bugs waiting to happen - if fields of the struct are reshuffled, for example.
use designated or, if initialized with zeroes, just omit the initializer body (or use `{0} in C`).

Reply:
Normally I use 

RGBW { .r = 1.0f, .g = 1.0f, .b = 10f, .w = 1.0f };

this should be safe, but for zero {0} would be better I agree

Reply:
> RGBW rgbw{ .r = 1.0f, .g = 1.0f, .b = 10f, .w = 1.0f };

this would be perfectly fine

> RGBW rgbw{ 1.0f, 1.0f, 10f, 1.0f };

this is not due to reason mentioned above.

> RGBW rgbw{ .r = 0.0f, .g = 0.0f, .b = 0.0f, .w = 0.0f };

is safe but too verbose, `RGBW rgbw{};` is nice and brief.